Feature/59925 show user popover on the name

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/59925

# What are you trying to accomplish?
Enable the hover card not only on avatars, but also when hovering on names.

- [x] patch AvatarComponent so that names rendered there will display a hover card.
- [x] search application for usernames that show up outside of the AvatarComponent
  - [x] linked names in WP journal
  - [x] within auto completers
  - [x] linked names in WP description and comments
  - [x] news
  - [x] wiki
  - [x] admin: users
  - [x] project: members

**Notable technical exceptions** where the hover card will not render due to technical limitations of our spot modal usage:

- work package share dialog
- global search
- global invitation modal

I am currently working on removing these limitations in a follow-up PR, but wanted to make the current state available for review.

**Notable other exceptions** where the required effort to patch in hover cards was not worth it in my opinion:

- the news block on the my-page: it uses a very different way of constructing links. I skipped it since the necessary increase in complexity is not worth it for a single widget.
- some username entries in WP journals that are not links, but plain text. This happens for example when the activity describes someone setting a new assignee on a work package. The new assignee will not be a link to a user, it will just be the username instead. Again skipped due to complexity vs. gain.

## Screenshots

https://github.com/user-attachments/assets/0a084df8-7bf5-4d58-b7c7-555ca5f787d5

# What approach did you choose and why?
I applied the same logic to usernames as I did on the avatars. ~For some user names (e.g. within auto completers), there was no easy way to inject the necessary parameters into the data attributes of the user name span, so I instead introduced a new CSS class: `.op-hover-card--data-on-sibling-principal`
If a hover card is triggered, but the source element has this class, the hover card data is sourced from sibling elements if possible.~

To cover most other links, I extended `ObjectLinking#link_to_user` to add hover card info to the link.

Fixed some timing and consistency bugs while on it, too.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#59925] principal renderer shows hover card on name hover

EinLama committed:
[#59925] fix: do not allow user hovecard within wp hover card

Nesting them is not supported.

EinLama committed:
[#59925] show hover card in wp activity

EinLama committed:
[#59925] show hover card for user links

used in

* news
* wiki
* activity
* and more!

EinLama committed:
[#59925] shared helper for rendering primer user links

EinLama committed:
[#59925] hover card in news block

... on the overview page

EinLama committed:
[#59925] hover card in new users block

... on overview page

EinLama committed:
[#59925] hover card in admin user table

EinLama committed:
[#59925] refactor

EinLama committed:
[#59925] fix specs

EinLama committed:
[#59925] make text formatter work with hover cards

After rendering a comment or wp description, the hover card should
appear when you hover over a mentioned username.

EinLama committed:
[#59925] Revert "[#59925] hover card in news block"

This reverts commit e0953619e04efeb8345bc525bc6fb7356256d9ae.

EinLama committed:
[#59925] use designated label for different use case

my-news uses angular, here we are using rails -> do not mix up the
translation keys as they are used with different approaches in mind.

EinLama committed:
[#59925] adjust for new translation

EinLama committed:
[#59925] fix user login display in user list

EinLama committed:
[#59925] enable hover cards for auto completer labels

EinLama committed:
[#59925] use more jQuery

EinLama committed:
[#59925] fix issue with partial name highlighting

The autocompleter hover cards only worked when there was no
search highlighting enabled. As soon as there were results,
the child spans were transformed and you could not hover over
them anymore, so cards did not trigger.

Setting these to `inline-block` fixes that issue.

EinLama committed:
[#59925] ensure correct modal is closed

EinLama committed:
[#59925] fix spec

EinLama committed:
[#59925] Revert "[#59925] ensure correct modal is closed"

This reverts commit 4fdc18c3880fee1bfa8aacb3c6e58f41601a69dc.

EinLama committed:
[#59925] Ensure hover cards play nice with other modals

EinLama committed:
[#59925] Provide some comments

EinLama committed:
[#59925] fix faulty spec

* use cuprite
* get rid of unneeded empty grid setup (that is ignored)
* expect a location that exists

EinLama committed:
[#59925] ignore rubocop for this method

Yes, it has many conditions. But throwing in additional methods won't
help with reducing its complexity.

HDinger commented on frontend/src/app/core/setup/globals/global-listeners/hover-card-trigger.service.ts:
```diff
@@ -154,13 +164,50 @@ export class HoverCardTriggerService {
+    return el;
+  }
+
+  private findSiblingElementWithHoverCardData(el:HTMLElement) {
+    const sibling = jQuery(el).siblings('.op-principal').get(0);
+    if (sibling) {
+      return this.childElementWithHoverCardTrigger(sibling);
+    }
+
+    return null;
+  }
+
+  private childElementWithHoverCardTrigger(principal:Element) {
+    return principal.querySelector('.op-hover-card--preview-trigger');
+  }
+
+  private parseHoverCardUrl(el:HTMLElement) {
+    return el.getAttribute('data-hover-card-url');
+  }
```
You rely heavily on a certain structure for this to work, but that is not clear when you just see the class. So whenever somewhen changes the structure in the autocompleter and e.g wrap the principal in a div, that will not work any more without the person noticing it.

I don't really get why all of this is necessary. You wrote that you could not add the hoverCard information directly on the element but why not?

EinLama replied:
We had a call about this and Henriette suggested a great way to overcome the obstacles I found while trying to place hovercard data directly onto the span. With these suggestions, all of the complicated methods above could be removed 🎉

HDinger reviewed (changes requested):
Hi @EinLama 

I don't quite get the necessity for the sibling logic. Maybe we can shortly talk about that. Other than that, I found some issues while testing it:

- [ ]  I managed quite often to get to a state where either the name or the avatar does not trigger a hoverCard any more
![Jan-31-2025 09-07-10](https://github.com/user-attachments/assets/1c2aad8b-c3ec-4044-b87c-69505ec4e76c)

- [ ] There seems to an issue with the helper, but I could not quite get my finger on it. The WP info line shortly shows the author as a link before it switches to a text. That does not happen with all users however, I only saw it when the author is my current logged in user. However, even when there is a link shown, there is no hoverCard when hovering it.

![Jan-31-2025 09-02-19](https://github.com/user-attachments/assets/f85e1dc5-9478-4262-96d1-67cdd48d22bd)

- [x] In the acceptance criteria it is said, that there should be no flicker when switching between name and avatar. The card is however disappearing when I leave the name and reopens when hovering the avatar. Due to the delay it does not feel flickering but I'd neverless like to hear the opinion of the product team on this.
![Jan-31-2025 09-04-03](https://github.com/user-attachments/assets/c9e9a4e2-fd86-4d8d-bfe9-fdf42654eb98)

EinLama committed:
[#59925] Get hover card data directly from username

EinLama committed:
[#59925] allow hover card to open even when mouse cursor is jittering

EinLama commented:
Hey @HDinger,

thank you for your review. You already helped me get rid of the unnecessary bloat fetching data from a sibling element. So that one is done.

> I managed quite often to get to a state where either the name or the avatar does not trigger a hoverCard any more

I could reproduce this by "jittering" the mouse cursor over the trigger. The hover card trigger service has a safe-guard that - once shown - will not re-show the same hover card again. This safe-guard was too restrictive and did not consider that you could move the mouse cursor away before the hover card is actually shown. So the trigger service thought it had to safe guard a hover card that wasn't there yet. I moved the safe-guard to the appropriate position and that fixed it. If the cause for your observed behavior was the same, that should be fixed, too.

>  There seems to an issue with the helper, but I could not quite get my finger on it. The WP info line shortly shows the author as a link before it switches to a text. That does not happen with all users however, I only saw it when the author is my current logged in user. However, even when there is a link shown, there is no hoverCard when hovering it.

I could not reproduce this, nor have I ever seen it 🤔 

>  In the acceptance criteria it is said, that there should be no flicker when switching between name and avatar. The card is however disappearing when I leave the name and reopens when hovering the avatar. Due to the delay it does not feel flickering but I'd neverless like to hear the opinion of the product team on this.

Will discuss later today and report back.

EinLama commented:
The hover card disappearing briefly when alternating between username and avatar [has been discussed ](https://community.openproject.org/meetings/4251#item-9333)with product and is deemed okay 👍🏻

HDinger commented on frontend/src/app/shared/components/autocompleter/user-autocompleter/user-autocompleter-template.component.sass:
```diff
@@ -11,3 +11,8 @@
+// For hover cards to work within the auto completer, we have to ensure that the hover event can be properly
+// triggered when mousing over the name. The easiest way to do this is to set it to inline-block.
+.op-user-autocompleter--name.op-hover-card--preview-trigger
+  display: inline-block
```
I don't quite understand why this is needed at all but it causes problems with long names which are simply cut off:

<img width="339" alt="Bildschirmfoto 2025-02-04 um 08 59 32" src="https://github.com/user-attachments/assets/8603144a-44d3-4b8a-854f-9d2395099804" />

EinLama replied:
I added this property since I encountered weird behavior. When you hovered over a name in an autocompleter, the hover card would show up. But as soon as you traced over a space between words, it would disappear - only to re-appear shortly thereafter when you leave the space and enter the next word.

However, this no longer happens now. So there must have been another in-development-state present back then that was the cause for this buggy behavior. Thanks for questioning this, I can remove this hack!

HDinger reviewed (changes requested):
I only have small remarks left:

* The author of a WP is now correctly shown as link (in the little info row next to the status). However, hovering over it does not trigger a HoverCard
* When opening a user autocompleter (e.g to set an assignee), I can also see groups. Hovering over a group name triggers an empty hovercard which lays behind the dropdown. There should be no card at all

<img width="607" alt="Bildschirmfoto 2025-02-04 um 08 58 47" src="https://github.com/user-attachments/assets/2329012b-d1ff-4228-b98c-684a98154388" />

EinLama committed:
[#59925] show hover card for work package author

EinLama committed:
[#59925] remove css hack

this is no longer needed

EinLama committed:
[#59925] autocompleters: only show hover card for users

EinLama commented:
> The author of a WP is now correctly shown as link (in the little info row next to the status). However, hovering over it does not trigger a HoverCard

Thank you, fixed!

> When opening a user autocompleter (e.g to set an assignee), I can also see groups. Hovering over a group name triggers an empty hovercard which lays behind the dropdown. There should be no card at all

That was a side-efffect of me removing the sibling logic. Fixed now so that hover cards are only shown for principals of type user.

EinLama committed:
[#59925] spec was fixed upstream, make green again

EinLama commented:
The red spec is flaky, have seen this a couple of times 😵

HDinger reviewed (approved):
Nice job 👍 🌟
